[InstCombine] Do not simplify lshr/shl arg if it is part of fshl rotate pattern.

quic-eikansh opened the pull request:
The fshl/fshr having first two arguments as same gets lowered to targets
specific rotate. But based on the uses, one of the arguments can get
simplified resulting in different arguments performing equivalent operation.

This patch prevents the simplification of the arguments of lshr/shl if they are
part of fshl pattern.

quic-eikansh committed:
[InstCombine] Refactoring matchFunnelShift (NFC)

The matchFunnelShift function was doing pattern matching and creating
the fshl/fshr instruction if needed. Moved the pattern matching code to function convertShlOrLShrToFShlOrFShr. It can be reused for other optimizations.

quic-eikansh committed:
Update formatting in InstCombineInternal.h

quic-eikansh committed:
[InstCombine] Added tests to fsh.ll

quic-eikansh commented:
This pull request is created to address the reviews in https://github.com/llvm/llvm-project/pull/66115

goldsteinn commented:
I would expect to see test difference in the commit adding the new restriction.

quic-eikansh commented:
> I would expect to see test difference in the commit adding the new restriction.

I didn't get what you meant by this. Should I have code changes and test in the same commit?

goldsteinn commented:
> > I would expect to see test difference in the commit adding the new restriction.
> 
> I didn't get what you meant by this. Should I have code changes and test in the same commit?

Yes, otherwise the commit with the impl will not be passing (and it makes review more difficult as its hard to see the proper side affects of a change).

quic-eikansh commented:
I have squashed the changes and test into one commit. @goldsteinn

goldsteinn commented:
> I have squashed the changes and test into one commit. @goldsteinn

Thats not really ideal either... Then its difficult to review the test changes.
Typically the way we do it is two seperate commits i.e

commit 1: Tests (w/ checks before the implementation)
commit 2: Implementation (re update test checks)

That way in commit 2 we can see what actual changes the implementation is causing.

quic-eikansh committed:
[InstCombine] Do not simplify lshr/shl arg if it is part of fshl rotate pattern.

The fshl/fshr having first two arguments as same gets lowered to targets
specific rotate. But based on the uses, one of the arguments can get
simplified resulting in different arguments performing equivalent
operation.

This patch prevents the simplification of the arguments of lshr/shl if
they are part of fshl pattern.

quic-eikansh commented:
@goldsteinn I have made the changes as you suggested.

quic-eikansh commented:
@goldsteinn Did you get a chance to look at it?

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:
```diff
@@ -2706,17 +2706,16 @@ Instruction *InstCombinerImpl::matchBSwapOrBitReverse(Instruction &I,
-                                     const DominatorTree &DT) {
+std::optional<std::tuple<Intrinsic::ID, SmallVector<Value *, 3>>>
+InstCombinerImpl::convertShlOrLShrToFShlOrFShr(Instruction &Or) {
   // TODO: Can we reduce the code duplication between this and the related
```
```suggestion
  assert(Or.getOpcode() == BinaryOperator::Or && "Expecting or instruction");
  // TODO: Can we reduce the code duplication between this and the related
```

quic-eikansh committed:
Removed DT from matchFunnelShift and added assert.

quic-eikansh commented on llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:
```diff
@@ -2706,17 +2706,16 @@ Instruction *InstCombinerImpl::matchBSwapOrBitReverse(Instruction &I,
-/// Match UB-safe variants of the funnel shift intrinsic.
-static Instruction *matchFunnelShift(Instruction &Or, InstCombinerImpl &IC,
-                                     const DominatorTree &DT) {
+std::optional<std::tuple<Intrinsic::ID, SmallVector<Value *, 3>>>
```
Do `std::pair`  has advantage over `std::tuple`? I see `std::tuple` used in codebase even for 2 element. I have addressed other 2 reviews.

nikic replied:
I agree that we should use std::pair instead of std::tuple if there are only two elements.

dtcxzyw reviewed (approved):
LGTM. Thanks!
Please wait for additional approval from other reviewers.

quic-eikansh commented:
@goldsteinn @nikic

quic-eikansh commented:
> LGTM. Thanks! Please wait for additional approval from other reviewers.

Thanks

dtcxzyw commented:
Ping.

efriedma-quic reviewed (approved):
Changes LGTM... but I'd like to land as a series of three commits; one for the refactor (6bbc776 and the fixup), one for the new tests (b896ff5), then one for the actual behavior change.

Please push separate PRs for the refactor and the test, then I'll merge everything for you.
